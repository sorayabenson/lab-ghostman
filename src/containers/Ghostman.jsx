import React, { Component } from 'react';
import '../components/style.css';
import Spinner from '../components/Spinner';
import Controls from '../components/Controls';
import { getApi, postApi, putApi, deleteApi } from '../services/api-utils';


export default class Ghostman extends Component {
    state = {
        loading: true,
        url: '',
        method: 'get',
        json: '',
        results: '',
        history: [],
    }

    componentDidMount() {
        this.setState({ loading: false })
    }

    handleUrlChange = (e) => {
        this.setState({ url: e.target.value }) 
    }

    handleMethodChange = (e) => {
        this.setState({ method: e.target.value })
        // this.setState({ method: e.target.value }, () => console.log(this.state.method))
        //calling console.log after this.setState doesn't work because it is dependent state, so you must call it as a call back within this.setState to see the real update.
    }

    handleJsonChange = (e) => {
        this.setState({ json: e.target.value })
    }

    refreshResults = () => {
        console.log(this.state.method, this.state.results)
    }

    handleSubmit = async (e) => {
        e.preventDefault();
        
        const { url, json } = this.state;
        this.setState({ loading: true });
        
        //if statement determines by method which service to send the url and json to
        if (this.state.method === 'get') {
            let results = await getApi(url);
            this.setState({ results: results });
        } else if (this.state.method === 'post') {
            let results = await postApi(url, json);
            this.setState({ results: results });
        } else if (this.state.method === 'put') {
            let results = await putApi(url, json);
            this.setState({ results: results });
        } else {
            let results = await deleteApi(url);
            this.setState({ results: results });
        }

        this.setState({ loading: false });
        
        //updates results
        this.refreshResults();    
        
        //makes an object of method, url, json and pushes it to history array

        //updates history
    }

    render() {
        const { loading, url,  json, results, history } = this.state;

        if (loading) return <Spinner />

        return (
            <>
                <Controls 
                    url={url}
                    json={json}
                    onUrlChange={this.handleUrlChange}
                    onMethodChange={this.handleMethodChange}
                    onJsonChange={this.handleJsonChange}
                    onSubmit={this.handleSubmit}
                />
            </>
        )
    }
}
