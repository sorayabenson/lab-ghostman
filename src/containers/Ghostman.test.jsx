import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Ghostman from './Ghostman';

describe('Ghostman component', () => {
  
  it('renders Ghostman', () => {
    render(<Ghostman />);
    
    const url = screen.getByRole('textbox', { name: 'api-url' })
    userEvent.type(url, 'https://personal-ghosts.herokuapp.com/ghosts/')
    expect(url).toMatchSnapshot();

    const getRadio = screen.getByRole('radio', { name: 'get' })
    expect(getRadio).toMatchSnapshot();
    userEvent.click(getRadio);
    expect(getRadio).toBeChecked();

    const postRadio = screen.getByRole('radio', { name: 'post' })
    expect(postRadio).toMatchSnapshot();
    userEvent.click(postRadio);
    expect(postRadio).toBeChecked();

    const putRadio = screen.getByRole('radio', { name: 'put' })
    expect(putRadio).toMatchSnapshot();
    userEvent.click(putRadio);
    expect(putRadio).toBeChecked();

    const deleteRadio = screen.getByRole('radio', { name: 'delete' })
    expect(deleteRadio).toMatchSnapshot();
    userEvent.click(deleteRadio);
    expect(deleteRadio).toBeChecked();

    const jsonText = screen.getByRole('textbox', { name: 'json-input' })
    userEvent.type(jsonText, 'yikes!')
    expect(jsonText).toMatchSnapshot();

    const button = screen.getByRole('button', { name: 'submit' })
    expect(button).toMatchSnapshot();
    userEvent.type(url, 'https://personal-ghosts.herokuapp.com/ghosts/1/')
    //getRoute test
    userEvent.click(getRadio)
    userEvent.click(button);
    screen.getByText('your ghosts will be with you shortly');
    //postRoute 
    // const ghost = '{\n  "name": "Test Ghost",\n "img": "http://placekitten.com/g/200/200",\n  "description": "test test",\n "category_id": 1,\n "price": 8,\n "price_currency": "tests",\n  "trustworthy": false\n}'
    // userEvent.type(url, 'https://personal-ghosts.herokuapp.com/ghosts/')
    // userEvent.type(jsonText, )
    // userEvent.click(postRadio)
    // userEvent.click(button);
    //putRoute test

    //deleteRoute test
    
  });
});
