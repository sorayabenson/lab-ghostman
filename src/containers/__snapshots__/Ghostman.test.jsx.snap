// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ghostman component renders Ghostman 1`] = `
<input
  aria-label="api-url"
  class="url"
  placeholder="http://type-api-url-here.com"
  type="url"
  value="https://personal-ghosts.herokuapp.com/ghosts/"
/>
`;

exports[`Ghostman component renders Ghostman 2`] = `
<input
  checked=""
  id="get"
  name="method"
  type="radio"
  value="get"
/>
`;

exports[`Ghostman component renders Ghostman 3`] = `
<input
  id="post"
  name="method"
  type="radio"
  value="post"
/>
`;

exports[`Ghostman component renders Ghostman 4`] = `
<input
  id="put"
  name="method"
  type="radio"
  value="put"
/>
`;

exports[`Ghostman component renders Ghostman 5`] = `
<input
  id="delete"
  name="method"
  type="radio"
  value="delete"
/>
`;

exports[`Ghostman component renders Ghostman 6`] = `
<textarea
  aria-label="json-input"
  class="jsonText"
  placeholder="{ 
   \\"send\\": \\"raw\\", 
   \\"json\\": \\"here\\" 
}"
>
  yikes!
</textarea>
`;

exports[`Ghostman component renders Ghostman 7`] = `
<pre
  class="resultsText"
>
  {}
</pre>
`;

exports[`Ghostman component renders Ghostman 8`] = `
<ul
  aria-label="history"
/>
`;

exports[`Ghostman component renders Ghostman 9`] = `
<button
  aria-label="submit"
>
  submit
</button>
`;
